Try to speedup update memory state

DanielPechersky opened the pull request:
First commit is a refactor for code clarity
Second commit is my attempted speedup, assuming `card.set_memory_state` is much slower than everything else.
My change will allocate an extra 3.51 MB of scratch space for every 10,000 cards the function processes.

DanielPechersky commented:
TODO: I don't update progress correctly

DanielPechersky commented:
@user1823 true, I got the impression off the discord that the FSRS code is the "known criminal" here, but there's no benchmark so the function could be slow for IO reasons. Obviously batching IO is a good idea too. 
Btw 10,000,000 cards = 351 MB. Not much, but the workaround is simple: Batch cards while processing them. I find it unlikely that the memory usage will be a problem though.

asukaminato0721 commented on rslib/src/scheduler/fsrs/memory_state.rs:
```diff
@@ -101,100 +99,122 @@ impl Collection {
+            for (mut card, original, _) in to_update_memory_state {
+                'reschedule_card: {
+                    // if rescheduling
+                    let Some(reviews) = &last_revlog_info else {
```
How about let-chains? 

https://rust-lang.github.io/rfcs/2497-if-let-chains.html

DanielPechersky replied:
Sure

DanielPechersky commented:
@Expertium Truly, my zero counting skills are legendary. You're right

L-M-Sherlock commented on rslib/src/scheduler/fsrs/memory_state.rs:
```diff
@@ -58,6 +58,18 @@ pub(crate) struct UpdateMemoryStateEntry {
+    fn chunk_into_vecs(&mut self, chunk_size: usize) -> impl Iterator<Item = Vec<T>>;
+}
+
+impl<T> ChunkIntoVecs<T> for Vec<T> {
+    fn chunk_into_vecs(&mut self, chunk_size: usize) -> impl Iterator<Item = Vec<T>> {
+        std::iter::from_fn(move || {
+            (!self.is_empty()).then(|| self.split_off(chunk_size.min(self.len())))
+        })
+    }
+}
```
The `split_off(self.len())` will return empty vecs.

L-M-Sherlock commented on rslib/src/scheduler/fsrs/memory_state.rs:
```diff
@@ -58,6 +58,18 @@ pub(crate) struct UpdateMemoryStateEntry {
+impl<T> ChunkIntoVecs<T> for Vec<T> {
+    fn chunk_into_vecs(&mut self, chunk_size: usize) -> impl Iterator<Item = Vec<T>> {
+        std::iter::from_fn(move || {
+            (!self.is_empty()).then(|| self.split_off(chunk_size.min(self.len())))
```
```suggestion
            (!self.is_empty()).then(|| self.drain(..chunk_size.min(self.len())).collect())
```

L-M-Sherlock commented:
It's nice to combine the progress. Here is my modification:

```diff
--- a/rslib/src/scheduler/fsrs/memory_state.rs
+++ b/rslib/src/scheduler/fsrs/memory_state.rs
@@ -17,2 +17,3 @@ use crate::card::CardType;
 use crate::prelude::*;
+use crate::progress::ThrottlingProgressHandler;
 use crate::revlog::RevlogEntry;
@@ -67,3 +68,3 @@ impl<T> ChunkIntoVecs<T> for Vec<T> {
         std::iter::from_fn(move || {
-            (!self.is_empty()).then(|| self.split_off(chunk_size.min(self.len())))
+            (!self.is_empty()).then(|| self.drain(..chunk_size.min(self.len())).collect())
         })
@@ -118,2 +119,5 @@ impl Collection {
 
+            let mut progress = self.new_progress_handler::<ComputeMemoryProgress>();
+            progress.update(false, |s| s.total_cards = items.len() as u32)?;
+
             let (items, cards_without_items): (Vec<(CardId, FsrsItemForMemoryState)>, Vec<CardId>) =
@@ -148,2 +152,3 @@ impl Collection {
                 usn,
+                &mut progress,
             )?;
@@ -230,2 +235,3 @@ impl Collection {
                 usn,
+                &mut progress,
             )?;
@@ -257,5 +263,4 @@ impl Collection {
         usn: Usn,
+        progress: &mut ThrottlingProgressHandler<ComputeMemoryProgress>,
     ) -> Result<()> {
-        let mut progress = self.new_progress_handler::<ComputeMemoryProgress>();
-        progress.update(false, |s| s.total_cards = cards.len() as u32)?;
         for (idx, card_id) in cards.into_iter().enumerate() {
@@ -278,2 +283,3 @@ impl Collection {
         usn: Usn,
+        progress: &mut ThrottlingProgressHandler<ComputeMemoryProgress>,
     ) -> Result<()> {
@@ -285,5 +291,2 @@ impl Collection {
         let mut starting_states = Vec::new();
-
-        let mut progress = self.new_progress_handler::<ComputeMemoryProgress>();
-        progress.update(false, |s| s.total_cards = items.len() as u32)?;
         for (i, items) in items.chunk_into_vecs(ITEM_CHUNK_SIZE).enumerate() {
```

DanielPechersky commented:
Looks good, feel free to commit that

DanielPechersky commented on rslib/src/scheduler/fsrs/memory_state.rs:
```diff
@@ -79,124 +92,247 @@ impl Collection {
-                None
+
+            let Some(req) = &req else {
+                let items = fsrs_items_for_memory_states(
```
I don't have the expertise to answer that, but the previous code behaves like this. Perhaps you can try this change in a separate PR, after this is merged?

DanielPechersky replied:
Yeah, the code is now easier to optimize in the 3 special-cased tight loops, but out of scope for this PR

DanielPechersky commented:
Do not merge: waiting on fsrs-rs 5.2.0

L-M-Sherlock commented on rslib/src/scheduler/fsrs/memory_state.rs:
```diff
@@ -79,124 +92,247 @@ impl Collection {
+                &mut on_updated_card,
+            )?;
+
+            let mut rescheduler = self
+                .get_config_bool(BoolKey::LoadBalancerEnabled)
+                .then(|| Rescheduler::new(self))
+                .transpose()?;
```
```suggestion
            let mut rescheduler = req
                .reschedule
                .bitand(self.get_config_bool(BoolKey::LoadBalancerEnabled))
                .then(|| Rescheduler::new(self))
                .transpose()?;
```

L-M-Sherlock replied:
We can skip `Rescheduler::new(self)` if `req.reschedule` is `false`. It could save ~60ms in a large collection.

L-M-Sherlock commented:
Oh, sorry... `bitand` is too fancy. `&&` is enough.

```rust
            let mut rescheduler = (req.reschedule
                && self.get_config_bool(BoolKey::LoadBalancerEnabled))
            .then(|| Rescheduler::new(self))
            .transpose()?;
```

Expertium commented:
@dae please take a look at this PR as well

DanielPechersky commented:
@dae I have updated the description

DanielPechersky commented:
I merged main in to fix merge conflicts. You can try https://stackoverflow.com/a/26986320/6211863 to see the diff, or I can rebase and remove the latest commit so someone else can fix the merge conflicts on merging to main.

dae commented:
> It would still be nice for the PR description to describe the current state of things, and what sort of improvement this has made.

Still would appreciate some actual numbers on the performance difference. :-)

L-M-Sherlock commented:
> Still would appreciate some actual numbers on the performance difference. :-)

I copied these figures from discord:

25.09 took 10.182176084s

Sahl's PR took 969.745458ms

Sahl's PR plus my PR took 853.057208ms

The test collection has ~46k cards with ~760k reviews.

dae commented:
That's an impressive improvement - well done to you both.

I've looked through the code briefly and it looks well written. My only remaining issue is with the smoke test - I don't think it's very useful at the moment, as you're running it on an empty collection. Adding at least one card and checking that one or more properties have been updated post-run as expected would be good. deckconfig/update.rs:updating() might be useful as a reference.

DanielPechersky commented:
@dae Is there anything else I need to do for this PR? Would you like me to rebase for a cleaner diff?

DanielPechersky commented:
@dae
1. I deleted the smoke test.
2. I added a test that verifies that fsrs params are cleared when req: None. Other functionality remains untested (same as before this PR)
3. I replaced [].as_slice() with &[].
4. I fixed the merge conflicts.
Please merge this or let me know if you still have concerns.

dae commented:
No objections to a merge from me. The contributors file has unfortunately taken a conflict in the interim. Would you mind rebasing so we can get this merged in?

user1823 commented on rslib/src/scheduler/fsrs/memory_state.rs:
```diff
@@ -80,139 +93,265 @@ impl Collection {
+                        return Ok(());
+                    };
+                    // the card isn't in (re)learning and suspended
+                    if card.ctype == CardType::Review && card.queue != CardQueue::Suspended {
```
```suggestion
                    if card.ctype != CardType::Review || card.queue == CardQueue::Suspended {
```

DanielPechersky replied:
My PR doesn't introduce this change, it was present on main. Let's adjust this after the PR is merged?

user1823 commented on rslib/src/scheduler/fsrs/memory_state.rs:
```diff
@@ -181,8 +181,8 @@ impl Collection {
                     };
-                    // the card isn't in (re)learning and suspended
-                    if card.ctype == CardType::Review && card.queue != CardQueue::Suspended {
+                    // the card is in relearning and it's not suspended
```
The comment was correct earlier.

```suggestion
                    // the card isn't in (re)learning or suspended
```

DanielPechersky replied:
The comments in this section are describing the reason we're continuing in this block of code rather than exiting early.

user1823 replied:
Yes. Here, we continue if the card is in review but we exit if the card is in (re)learning. So, the comment should read "the card isn't in (re)learning", right?

abdnh commented:
We were doing some changes to merge permissions and I can't merge PRs now for some reason. @andrewsanchez can you merge it?

<img width="1065" height="160" alt="image" src="https://github.com/user-attachments/assets/37f45880-c2f1-4c54-87d1-1a8980d19252" />
